addons/l10n_sa_edi/models/account_move.py

```diff
@@ -0,0 +1,171 @@
+                                                                                                           PublicFormat.SubjectPublicKeyInfo))
+            qr_code_str = (seller_name_enc + seller_vat_enc + timestamp_enc + amount_total_enc +
+                           amount_tax_enc + invoice_hash_enc + signature_enc + public_key_enc)
+        return qr_code_str
+
+    def _l10n_sa_get_previous_invoice(self):
+        """
+            Get the previously submitted EDI document, whether it was accepted or rejected.
+        """
+        return self.env['l10n_sa_edi.submission'].search(
+            [('l10n_sa_journal_id', '=', self.journal_id.id), ('l10n_sa_attachment_id', '!=', False)], limit=1)
+
+    @api.depends('state', 'edi_document_ids.state')
+    def _compute_edi_show_cancel_button(self):
+        """
+            Override to hide the EDI Cancellation button at all times for ZATCA Invoices
+        """
+        super()._compute_edi_show_cancel_button()
+        for move in self.filtered(lambda m: move.is_invoice() and move.country_code == 'SA'):
+            move.edi_show_cancel_button = False
+
+    def button_recalculate_edi_values(self):
+        """
+            Force recalculate unsigned data (UUID, XML content, signature) for the invoice in case it is needed.
+        """
+        return self._l10n_sa_generate_unsigned_data()
+
+    def _l10n_sa_generate_unsigned_data(self):
+        """
+            Generate Unsigned XML content & digital signature to be used during both Signing and QR code generation.
+            It is necessary to save the signature as it changes everytime it is generated and both the signing and the
+            QR code expect to have the same, identical signature.
+        """
+        self.ensure_one()
+        edi_format = self.env.ref('l10n_sa_edi.edi_sa_zatca')
+        # Build the dict of values to be used for generating the Invoice XML content
+        invoice_values = edi_format._l10n_sa_prepare_values(self)
+        # Set Invoice field values required for generating the XML content, hash and signature
+        self.l10n_sa_uuid = uuid.uuid4()
+        self.l10n_sa_unsigned_xml = edi_format._l10n_sa_generate_zatca_template(self, invoice_values)
+        # Once the required values are generated, we hash the invoice, then use it to generate a Signature
+        invoice_hash_hex = edi_format._l10n_sa_generate_invoice_xml_hash(self.l10n_sa_unsigned_xml).decode()
+        self.l10n_sa_invoice_signature = edi_format._l10n_sa_get_digital_signature(self.journal_id.company_id,
+                                                                                   invoice_hash_hex).decode()
+
+    def _l10n_sa_save_submission(self, xml_content, invoice_name, response_data=None):
+        """
+            Save submitted invoice XML content in case of either Rejection or Acceptance.
+            This is useful to figure out why an invoice was rejected since we also save the errors along with the state.
```

Comment:
@jco-odoo Sometimes we receive warnings instead of errors, and Nada (SEN) expressed interest in seeing the warnings even in the case of successful submissions